x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java

```diff
@@ -310,6 +318,31 @@ private static void parseHit(SearchHit hit, String configId, RetrievedStatsHolde
         );
     }
 
+    private List<Stats> buildStatsWithDefaults(List<DataFrameAnalyticsConfig> configs, List<Stats> stats) {
+        Map<String, DataFrameAnalyticsConfig> configById = new HashMap<>();
+        for (DataFrameAnalyticsConfig config : configs) {
+            configById.put(config.getId(), config);
+        }
+
+        List<Stats> statsWithDefaults = new ArrayList<>(stats.size());
+        for (Stats statsItem : stats) {
+            DataFrameAnalyticsConfig config = configById.get(statsItem.getId());
+            statsWithDefaults.add(new Stats(
+                statsItem.getId(),
+                statsItem.getState(),
+                statsItem.getFailureReason(),
+                statsItem.getProgress(),
+                statsItem.getDataCounts() == null ? new DataCounts(config.getId()) : statsItem.getDataCounts(),
```

Comment:
Why can't this and the memory usage be initialized in the stats ctor? 
```
this.memoryUsage = memoryUsage == null ? new MemoryUsage(config.getId(), Instant.now(), 0) : memoryUsage;
this.dataCounts = dataCounts == null ? new DataCounts(config.getId()) : dataCounts;
```

Why do we need to set `memory_usage` create time to the config's create time? Is that adding value for the added complexity?

Or maybe I am misunderstanding the `Stats#getId()` value...

Reply:
I agree that's too much complexity for the gain. I'll make `MemoryUsage.timestamp` nullable.